{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import time\n",
    "import csv\n",
    "def Main():\n",
    "    print(\"\"\"Welcome to the Codon Frequency Optimizer! For easiest use, ensure that\n",
    "    you are running the program from its original folder! \n",
    "    ****NOTE: Make sure the sequence you are optimizing begins with the first\n",
    "    nucleotide and is in the correct reading frame!****\"\"\")\n",
    "\n",
    "    #Get and validify sequence\n",
    "    valid_mRNA = True\n",
    "    seq = input(\"Please input your DNA or mRNA sequence: \")\n",
    "    try:\n",
    "        seq_str = str(seq)\n",
    "    except: \n",
    "        print(\"Sequence has invalid characters. Terminating program.\")\n",
    "        return \"invalid\"\n",
    "    mRNA_seq = \"\"\n",
    "    \n",
    "    if ((seq_str.find(\"U\") != -1) or (seq_str.find(\"c\") != -1)\n",
    "        or (seq_str.find(\"a\") != -1) or (seq_str.find(\"g\") != -1)\n",
    "        or (seq_str.find(\"u\") != -1)):\n",
    "        for char in seq_str:\n",
    "            if ((char == \"U\") or (char ==\"u\")\n",
    "                or (char == \"t\")):\n",
    "                mRNA_seq += \"T\"\n",
    "            elif (char == \"g\"): mRNA_seq += \"G\"\n",
    "            elif (char == \"c\"): mRNA_seq += \"C\"\n",
    "            elif (char == \"a\"): mRNA_seq += \"A\"\n",
    "            else: mRNA_seq += char\n",
    "    else: mRNA_seq = seq #if no typing corrections\n",
    "    for char in mRNA_seq:\n",
    "        if ((char != \"A\") and (char != \"T\")\n",
    "            and (char != \"C\") and (char !=\"G\")):\n",
    "            valid_mRNA = False\n",
    "            print(\"WRONG\")\n",
    "    if valid_mRNA == True:\n",
    "        print(\"It's good\")\n",
    "        print(mRNA_seq)\n",
    "    else: \n",
    "        print(\"Sequence has invalid characters. Terminating program.\")\n",
    "        return \"invalid\"\n",
    "       \n",
    "    seq_start = mRNA_seq.find(\"ATG\")\n",
    "    if (seq_start == -1): \n",
    "        print(\"This sequence has no start codon. Continue?\")\n",
    "        seq_cont = input(\"[Y]es or [N]o: \" )\n",
    "        if ((seq_cont == \"Y\") or (seq_cont == \"y\")\n",
    "               or (seq_cont == \"N\") or (seq_cont == \"n\")):\n",
    "            cont = True\n",
    "        else: cont = False\n",
    "        while (cont == False):\n",
    "            seq_cont = input(\"Invalid response. Continue?\\n[Y]es or [N]o: \")\n",
    "            if ((seq_cont == \"Y\") or (seq_cont == \"y\")\n",
    "               or (seq_cont == \"N\") or (seq_cont == \"n\")):\n",
    "                cont = True\n",
    "                \n",
    "            print(str(cont))\n",
    "        if ((seq_cont == \"N\") or (seq_cont == \"n\")):\n",
    "            print(\"Terminating program.\")\n",
    "            return \"invalid\"\n",
    "        else: print(\"Proceeding.\")\n",
    "\n",
    "    #import host codon sequence\n",
    "    #sequences obtained from GenScript site\n",
    "    try: \n",
    "        print(\"Script executed from: \")\n",
    "        print(os.path.dirname(os.path.realpath(\"Frequency Codon Optimizer.ipynb\")))\n",
    "        path = os.path.dirname(os.path.realpath(\"Frequency Codon Optimizer.ipynb\"))\n",
    "        print(\"The path is: \" + path)\n",
    "    except:\n",
    "        valid_path = False\n",
    "        path = input(\"Couldn't find active directory for this script. Please input the active directory: \")\n",
    "        while (valid_path == False):\n",
    "            if (os.path.isdir(path) == False):\n",
    "                path = input(\"Input directory was invalid, please try again or enter 'Q' to quit: \")\n",
    "                if ((path == \"Q\") or (path == \"q\")): return \"invalid\"\n",
    "            else: valid_path = True\n",
    "    \n",
    "    print(\"Found directory. Moving Forward\")\n",
    "    print(path + \"\\\\Template_Organisms\")\n",
    "    if (os.path.isdir(path + \"\\\\Template_Organisms\") == True):\n",
    "        path_files = path + \"\\\\Template_Organisms\"\n",
    "        i = 1\n",
    "        dir_items = []\n",
    "        print(\"Found template folder, listing template files: \")\n",
    "        for file in os.listdir(path_files):\n",
    "            print(\"\\t\" + str(i) + \": \" + file)\n",
    "            i+=1\n",
    "            dir_items.append(file)\n",
    "        \n",
    "    else: \n",
    "        print(\"Template_Organisms folder is not at listed directory. Terminating program.\")\n",
    "        return \"invalid\"\n",
    "    valid_codon_freq = False\n",
    "    while (valid_codon_freq == False):\n",
    "        ##Select host organism\n",
    "        host_codon_sel = input(\"Please input the number that corresponds with the host organism or 'Q' to quit: \")\n",
    "        if ((host_codon_sel == \"Q\") or (host_codon_sel == \"q\")): return \"invalid\"\n",
    "        host_codon_sel_valid = False\n",
    "        while (host_codon_sel_valid == False):\n",
    "            try: \n",
    "                host_codon_int = int(host_codon_sel)\n",
    "                if ((host_codon_int < (len(dir_items) + 1)) and (host_codon_int > 0)):\n",
    "                    host_codon_sel_valid = True\n",
    "                else: \n",
    "                    host_codon_sel = input(\"\"\"Input was not within the list range. Please input the number that \n",
    "                    corresponds with the host organism or 'Q' to quit: \"\"\")\n",
    "            except: host_codon_sel = input(\"\"\"Input was not an integer. Please input the number that \n",
    "            corresponds with the host organism or 'Q' to quit: \"\"\")\n",
    "            if ((host_codon_sel== \"Q\") or (host_codon_sel == \"q\")): return \"invalid\"\n",
    "\n",
    "        print(\"Selected host organism file: \" + dir_items[host_codon_int-1])\n",
    "        host_codon = dir_items[host_codon_int-1]\n",
    "\n",
    "        ##Select Target codon frequency\n",
    "        targ_codon_sel = input(\"\"\"Please input the number that corresponds with the target organism \n",
    "        codon frequency or 'Q' to quit: \"\"\")\n",
    "        if ((targ_codon_sel == \"Q\") or (targ_codon_sel == \"q\")): return \"invalid\"\n",
    "        targ_codon_sel_valid = False\n",
    "        while (targ_codon_sel_valid == False):\n",
    "            try: \n",
    "                targ_codon_int = int(targ_codon_sel)\n",
    "                if ((targ_codon_int < (len(dir_items) + 1)) and (targ_codon_int > 0)):\n",
    "                    targ_codon_sel_valid = True\n",
    "                else: \n",
    "                    targ_codon_sel = input(\"\"\"Input was not within the list range. Please input the number that \n",
    "                    corresponds with the target organism codon frequency or 'Q' to quit: \"\"\")\n",
    "            except: targ_codon_sel = input(\"\"\"Input was not an integer. Please input the number that \n",
    "            corresponds with the target organism codon frequency or 'Q' to quit: \"\"\")\n",
    "            if ((targ_codon_sel== \"Q\") or (targ_codon_sel == \"q\")): return \"invalid\"\n",
    "    \n",
    "        print(\"Selected target organism file: \" + dir_items[targ_codon_int-1])\n",
    "        targ_codon = dir_items[targ_codon_int-1]\n",
    "        \n",
    "        if (targ_codon_int == host_codon_int): \n",
    "            print(\"The selected host and target organisms are the same.\\nPlease select 2 different organisms.\")\n",
    "        else: valid_codon_freq = True\n",
    "    \n",
    "    host_file_path = path_files + \"\\\\\" + dir_items[host_codon_int-1]\n",
    "    targ_file_path = path_files + \"\\\\\" + dir_items[targ_codon_int-1]\n",
    "    \n",
    "    #store file contents in parallel arrays\n",
    "    with open(host_file_path, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        host_codons = [row[\"Triplet\"]for row in reader]\n",
    "        x.close()\n",
    "    with open(host_file_path, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        host_aa = [row[\"Amino acid\"] for row in reader]\n",
    "        x.close()\n",
    "    with open(host_file_path, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        host_freqs = [row[\"Fraction\"] for row in reader]\n",
    "        x.close()\n",
    "    with open(targ_file_path, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        targ_codons = [row[\"Triplet\"]for row in reader]\n",
    "        x.close()\n",
    "    with open(targ_file_path, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        targ_aa = [row[\"Amino acid\"] for row in reader]\n",
    "        x.close()\n",
    "    with open(targ_file_path, \"rt\") as x:\n",
    "        reader = csv.DictReader(x)\n",
    "        targ_freqs = [row[\"Fraction\"] for row in reader]\n",
    "        x.close()\n",
    "        \n",
    "    print(host_codons,host_aa, host_freqs)\n",
    "    print(targ_codons,targ_aa,targ_freqs)\n",
    "    \n",
    "    ## convert original sequence to AA and store frequencies\n",
    "    seq_len = len(mRNA_seq)\n",
    "    curr_codon = \"\"\n",
    "    new_aa_seq = \"\"\n",
    "    for i in range(0,seq_len-1):\n",
    "        curr_codon += mRNA_seq[i]\n",
    "        if (len(curr_codon) == 3):\n",
    "            for index, j in enumerate(host_codons):\n",
    "                if curr_codon == j:\n",
    "                    new_aa_seq += host_aa[index]\n",
    "                    break;\n",
    "                    \n",
    "            curr_codon = \"\"\n",
    "    print(\"The current seq is: \" + new_aa_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Welcome to the Codon Frequency Optimizer! For easiest use, ensure that\n",
      "    you are running the program from its original folder! \n",
      "    ****NOTE: Make sure the sequence you are optimizing begins with the first\n",
      "    nucleotide and is in the correct reading frame!****\n",
      "Please input your DNA or mRNA sequence: ATGATGATGATGATGATGATGAA\n",
      "It's good\n",
      "ATGATGATGATGATGATGATGAA\n",
      "This sequence has no start codon. Continue?\n",
      "[Y]es or [N]o: Y\n",
      "Proceeding.\n",
      "Script executed from: \n",
      "C:\\Users\\tlorg\\Documents\\GitHub\\Frequency_Codon_Optimizer\n",
      "The path is: C:\\Users\\tlorg\\Documents\\GitHub\\Frequency_Codon_Optimizer\n",
      "Found directory. Moving Forward\n",
      "C:\\Users\\tlorg\\Documents\\GitHub\\Frequency_Codon_Optimizer\\Template_Organisms\n",
      "Found template folder, listing template files: \n",
      "\t1: Arabidopsis_thaliana.csv\n",
      "\t2: Caenorhabditis_elegans.csv\n",
      "\t3: Cricetulus_griseus.csv\n",
      "\t4: Drosophila_melanogaster.csv\n",
      "\t5: Escherichia_coli.csv\n",
      "\t6: Human.csv\n",
      "\t7: Insect.csv\n",
      "\t8: Mouse.csv\n",
      "\t9: Nicotiana_tabacum.csv\n",
      "\t10: Pichia_pastoris.csv\n",
      "\t11: Pig.csv\n",
      "\t12: Rat.csv\n",
      "\t13: Saccharomyces_cerevisiae.csv\n",
      "\t14: Streptomyces.csv\n",
      "\t15: Yeast.csv\n",
      "\t16: Zae_mays.csv\n",
      "Please input the number that corresponds with the host organism or 'Q' to quit: 10\n",
      "Selected host organism file: Pichia_pastoris.csv\n",
      "Please input the number that corresponds with the target organism \n",
      "        codon frequency or 'Q' to quit: 6\n",
      "Selected target organism file: Human.csv\n",
      "['TTT', 'TTC', 'TTA', 'TTG', 'TAT', 'TAC', 'TAA', 'TAG', 'CTT', 'CTC', 'CTA', 'CTG', 'CAT', 'CAC', 'CAA', 'CAG', 'ATT', 'ATC', 'ATA', 'ATG', 'AAT', 'AAC', 'AAA', 'AAG', 'GTT', 'GTC', 'GTA', 'GTG', 'GAT', 'GAC', 'GAA', 'GAG', 'TCT', 'TCC', 'TCA', 'TCG', 'TGT', 'TGC', 'TGA', 'TGG', 'CCT', 'CCC', 'CCA', 'CCG', 'CGT', 'CGC', 'CGA', 'CGG', 'ACT', 'ACC', 'ACA', 'ACG', 'AGT', 'AGC', 'AGA', 'AGG', 'GCT', 'GCC', 'GCA', 'GCG', 'GGT', 'GGC', 'GGA', 'GGG'] ['F', 'F', 'L', 'L', 'Y', 'Y', '*', '*', 'L', 'L', 'L', 'L', 'H', 'H', 'Q', 'Q', 'I', 'I', 'I', 'M', 'N', 'N', 'K', 'K', 'V', 'V', 'V', 'V', 'D', 'D', 'E', 'E', 'S', 'S', 'S', 'S', 'C', 'C', '*', 'W', 'P', 'P', 'P', 'P', 'R', 'R', 'R', 'R', 'T', 'T', 'T', 'T', 'S', 'S', 'R', 'R', 'A', 'A', 'A', 'A', 'G', 'G', 'G', 'G'] ['0.56', '0.44', '0.15', '0.33', '0.45', '0.55', '0.53', '0.29', '0.17', '0.08', '0.12', '0.16', '0.54', '0.46', '0.62', '0.38', '0.51', '0.31', '0.18', '1', '0.48', '0.52', '0.47', '0.53', '0.42', '0.23', '0.16', '0.2', '0.59', '0.41', '0.58', '0.42', '0.29', '0.2', '0.19', '0.09', '0.65', '0.35', '0.18', '1', '0.35', '0.16', '0.4', '0.1', '0.16', '0.05', '0.11', '0.05', '0.4', '0.24', '0.25', '0.11', '0.15', '0.09', '0.47', '0.16', '0.45', '0.25', '0.24', '0.06', '0.43', '0.14', '0.32', '0.1']\n",
      "['TTT', 'TTC', 'TTA', 'TTG', 'TAT', 'TAC', 'TAA', 'TAG', 'CTT', 'CTC', 'CTA', 'CTG', 'CAT', 'CAC', 'CAA', 'CAG', 'ATT', 'ATC', 'ATA', 'ATG', 'AAT', 'AAC', 'AAA', 'AAG', 'GTT', 'GTC', 'GTA', 'GTG', 'GAT', 'GAC', 'GAA', 'GAG', 'TCT', 'TCC', 'TCA', 'TCG', 'TGT', 'TGC', 'TGA', 'TGG', 'CCT', 'CCC', 'CCA', 'CCG', 'CGT', 'CGC', 'CGA', 'CGG', 'ACT', 'ACC', 'ACA', 'ACG', 'AGT', 'AGC', 'AGA', 'AGG', 'GCT', 'GCC', 'GCA', 'GCG', 'GGT', 'GGC', 'GGA', 'GGG'] ['F', 'F', 'L', 'L', 'Y', 'Y', '*', '*', 'L', 'L', 'L', 'L', 'H', 'H', 'Q', 'Q', 'I', 'I', 'I', 'M', 'N', 'N', 'K', 'K', 'V', 'V', 'V', 'V', 'D', 'D', 'E', 'E', 'S', 'S', 'S', 'S', 'C', 'C', '*', 'W', 'P', 'P', 'P', 'P', 'R', 'R', 'R', 'R', 'T', 'T', 'T', 'T', 'S', 'S', 'R', 'R', 'A', 'A', 'A', 'A', 'G', 'G', 'G', 'G'] ['0.45', '0.55', '0.07', '0.13', '0.43', '0.57', '0.28', '0.2', '0.13', '0.2', '0.07', '0.41', '0.41', '0.59', '0.25', '0.75', '0.36', '0.48', '0.16', '1', '0.46', '0.54', '0.42', '0.58', '0.18', '0.24', '0.11', '0.47', '0.46', '0.54', '0.42', '0.58', '0.18', '0.22', '0.15', '0.06', '0.45', '0.55', '0.52', '1', '0.28', '0.33', '0.27', '0.11', '0.08', '0.19', '0.11', '0.21', '0.24', '0.36', '0.28', '0.12', '0.15', '0.24', '0.2', '0.2', '0.26', '0.4', '0.23', '0.11', '0.16', '0.34', '0.25', '0.25']\n",
      "The current seq is: MMMMMMM\n"
     ]
    }
   ],
   "source": [
    "response = Main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
